### ISSUE TITLE
Propagation of errors from helper/leader to collector

### ISSUE THREAD

**Issue by tgeoghegan** [ietf-wg-ppm/draft-ietf-ppm-dap#57]
Client reports could be rejected by the leader and helper for any number of reasons (decryption failure, proof didn't check out, malformed data, etc.). Since client participation in the protocol ends after the upload phase, there is no way for the client to find out about it (even if we move as much validation as possible to the upload phase, I believe there will always be failures that can't be detected until verify). The protocol should provide a way for leaders and helpers to inform collectors of errors so that misbehaving clients can be fixed.

**Comment by tgeoghegan** [comment]
Some prior art here from Prio v2: https://github.com/abetterinternet/prio-server/issues/6. We never implemented this, except that a list of UUIDs of packets (packets are analogous to a report in Priov3) rejected due to bad proofs is included with the sum parts transmitted to the portal server.

**Comment by cjpatton** [comment]
A specific error we need to account for here is decryption failure due to a stale HPKE config. (This was discussed originally in #49.)

**Comment by simon-friedberger** [comment]
Side-note: As long as the aggregators can decrypt their shares they can run validation even if a report would not be included in a batch because it has expired. I doubt that it would be worth the effort but since Max was asking about detecting when validation is too restrictive on the mailing list it might make sense in some cases.

**Comment by cjpatton** [comment]
As of DAP-01 we have enumerated in the spec the reasons for which a report can be rejected, so it's may be worth dusting off this conversation.

@tgeoghegan what would be useful here? Do we need to change the wire protocol? What about a JSON blob in the HTTP response header for the CollectResp that includes some telemetry, e.g., the number of times each rejection type occurred for the batch?

**Comment by tgeoghegan** [comment]
I think if we do anything here, it would be to have the leader nest the helper's RFC 7807 problem document in the problem document that the leader presents to the collector in response to a request to a collect URI. So I think maybe we'd have to define a problem type like "helperRefusedRequest".

However, I also think we should avoid prescribing anything unless necessary. I think the only case where we might need to relay an error like this is if a collector polls a collect URI, which causes a leader to make an `AggregateShareReq` to the helper, and then the helper fails to service that request. So under what circumstances would that happen, and what would we want the collector to do about it?

If the `AggregateShareReq` fails because of something like the helper having crashed or being unavailable, then there's nothing for the collector to do but try again later, and the leader providing an HTTP 500 suffices for that, meaning we don't need to spend more protocol text spelling out a more specific error.

The other failure scenario that comes to mind is if a helper rejects an `AggregateShareReq` because of something like insufficient reports, or some other kind of privacy violation. But if that's the case, then wouldn't the leader have rejected the `CollectReq` before the helper got a chance to reject the `AggregateShareReq`? This scenario points to some kind of disagreement between the leader and helper on the state of the world, probably due to a bug in one or the other. But again, what's the collector meant to do here but try again later?

Long story short, I think we should say nothing until we come across a scenario where propagating an error actually informs the behavior of a protocol actor. The nice thing about RFC 7807 is that it [allows extensions to errors](https://www.rfc-editor.org/rfc/rfc7807#section-3.2), which means that implementations are free to construct complex error objects that provide whatever context they want. With that in mind, I'd support closing this issue.

**Comment by simon-friedberger** [comment]
This was mentioned in some other issue: Assuming that the Collector probably also provides the Client, it would be good to tell the Collector things like "the client is using the wrong HPKE config" or "I am receiving a lot of requests for your tasks which do not aggregate into any reports because of their metadata". I probably wouldn't prescribe this kind of UX issue on the protocol level, though.

**Comment by cjpatton** [comment]
@simon-friedberger something we might do in our own deployments is have the Aggregators expose something like Prometheus metrics to the Collector. Would that be sufficient?

Does anyone still think we need language in the DAP spec to cover this issue?

**Comment by tgeoghegan** [comment]
This issue has been around for a long time and I don't think anyone has made a compelling case for including aggregator to collector error reporting mechanisms in-band in DAP. I'm closing this issue as we don't plan to take action on it in DAP. Naturally this doesn't rule out implementations of DAP from setting up their own bespoke error reporting mechanisms.
